File: examples/node/src/grpc/error-handling.ts
/**
 * Error Handling in gRPC Streaming: Rate Limiting, Timeouts, and Validation
 *
 * This example demonstrates proper error handling patterns for gRPC streaming:
 * - Input validation with structured errors
 * - Rate limiting and resource exhaustion
 * - Timeouts and deadline exceeded
 * - Permission checks and authorization
 * - Mid-stream errors and recovery
 *
 * gRPC Status Codes Used:
 * - INVALID_ARGUMENT: Bad request data
 * - PERMISSION_DENIED: Authorization failure
 * - RESOURCE_EXHAUSTED: Rate limit exceeded
 * - DEADLINE_EXCEEDED: Operation timeout
 * - INTERNAL: Server-side errors
 * - UNAVAILABLE: Service temporarily unavailable
 *
 * @see https://grpc.io/docs/guides/error/
 * @see crates/spikard-node/src/grpc/handler.rs
 */

import type { GrpcRequest, GrpcResponse } from '@spikard/node';

// gRPC Status Codes
enum GrpcStatus {
  OK = 0,
  CANCELLED = 1,
  UNKNOWN = 2,
  INVALID_ARGUMENT = 3,
  DEADLINE_EXCEEDED = 4,
  NOT_FOUND = 5,
  ALREADY_EXISTS = 6,
  PERMISSION_DENIED = 7,
  RESOURCE_EXHAUSTED = 8,
  FAILED_PRECONDITION = 9,
  ABORTED = 10,
  OUT_OF_RANGE = 11,
  UNIMPLEMENTED = 12,
  INTERNAL = 13,
  UNAVAILABLE = 14,
  DATA_LOSS = 15,
  UNAUTHENTICATED = 16,
}

// Mock protobuf types
interface DataRequest {
  userId: string;
  query: string;
  limit?: number;
}

interface DataResponse {
  result: string;
  processedAt: number;
}

/**
 * Rate limiter implementation
 */
class RateLimiter {
  private requests: Map<string, number[]> = new Map();

  /**
   * Check if user exceeded rate limit
   *
   * @param userId User identifier
   * @param maxRequests Maximum requests allowed
   * @param windowMs Time window in milliseconds
   * @returns true if within limit, false if exceeded
   */
  checkLimit(userId: string, maxRequests: number, windowMs: number): boolean {
    const now = Date.now();
    const userRequests = this.requests.get(userId) ?? [];

    // Remove requests outside the time window
    const recentRequests = userRequests.filter((timestamp) => now - timestamp < windowMs);

    // Check if limit exceeded
    if (recentRequests.length >= maxRequests) {
      return false;
    }

    // Add current request
    recentRequests.push(now);
    this.requests.set(userId, recentRequests);

    return true;
  }

  /**
   * Get remaining requests for user
   */
  getRemainingRequests(userId: string, maxRequests: number, windowMs: number): number {
    const now = Date.now();
    const userRequests = this.requests.get(userId) ?? [];
    const recentRequests = userRequests.filter((timestamp) => now - timestamp < windowMs);

    return Math.max(0, maxRequests - recentRequests.length);
  }
}

/**
 * Data Service Handler with Comprehensive Error Handling
 */
class DataServiceHandler {
  private rateLimiter = new RateLimiter();
  private readonly MAX_REQUESTS_PER_MINUTE = 10;
  private readonly RATE_LIMIT_WINDOW_MS = 60000; // 1 minute
  private readonly MAX_QUERY_LENGTH = 1000;
  private readonly MAX_LIMIT = 100;

  /**
   * Unary RPC with error handling
   */
  async handleRequest(request: GrpcRequest): Promise<GrpcResponse> {
    console.log(`[Unary] ${request.methodName} called`);

    try {
      // 1. Validate authentication
      this.validateAuth(request);

      // Deserialize request
      const payload = JSON.parse(request.payload.toString('utf-8')) as DataRequest;

      // 2. Validate input
      this.validateInput(payload);

      // 3. Check rate limit
      this.checkRateLimit(payload.userId);

      // 4. Check permissions
      this.checkPermissions(payload.userId, payload.query);

      // 5. Process request (simulate work)
      await this.delay(100);

      // Success response
      const response: DataResponse = {
        result: `Processed query: ${payload.query}`,
        processedAt: Date.now(),
      };

      return {
        payload: Buffer.from(JSON.stringify(response), 'utf-8'),
        metadata: {
          'x-rate-limit-remaining': String(
            this.rateLimiter.getRemainingRequests(
              payload.userId,
              this.MAX_REQUESTS_PER_MINUTE,
              this.RATE_LIMIT_WINDOW_MS
            )
          ),
        },
      };
    } catch (error) {
      // Convert errors to gRPC errors
      throw this.toGrpcError(error);
    }
  }

  /**
   * Server Streaming RPC with mid-stream error handling
   */
  async handleServerStream(request: GrpcRequest): Promise<GrpcResponse> {
    console.log(`[Server Streaming] ${request.methodName} called`);

    const responses: DataResponse[] = [];

    try {
      // Initial validation
      this.validateAuth(request);

      const payload = JSON.parse(request.payload.toString('utf-8')) as DataRequest;
      this.validateInput(payload);
      this.checkRateLimit(payload.userId);

      const limit = payload.limit ?? 10;

      // Process stream with potential mid-stream errors
      for (let i = 0; i < limit; i++) {
        // Simulate mid-stream failure at halfway point
        if (i === Math.floor(limit / 2)) {
          console.log('  ⚠️  Simulating mid-stream error');
          throw new Error('UNAVAILABLE: Service temporarily unavailable');
        }

        const response: DataResponse = {
          result: `Result ${i + 1} for query: ${payload.query}`,
          processedAt: Date.now() + i * 100,
        };

        responses.push(response);
        console.log(`  Generated response ${i + 1}/${limit}`);

        await this.delay(50);
      }

      return {
        payload: Buffer.from(JSON.stringify({ responses }), 'utf-8'),
        metadata: {},
      };
    } catch (error) {
      // Mid-stream errors need special handling
      console.error('  ✗ Stream error:', error);
      throw this.toGrpcError(error);
    }
  }

  /**
   * Client Streaming RPC with batch validation
   */
  async handleClientStream(
    requestStream: AsyncIterator<GrpcRequest>
  ): Promise<GrpcResponse> {
    console.log('[Client Streaming] ProcessBatch called');

    const results: string[] = [];
    let requestCount = 0;
    const MAX_BATCH_SIZE = 50;

    try {
      for await (const request of requestStream) {
        requestCount++;

        // Enforce batch size limit
        if (requestCount > MAX_BATCH_SIZE) {
          throw new Error(
            `RESOURCE_EXHAUSTED: Batch size exceeded (max ${MAX_BATCH_SIZE})`
          );
        }

        // Validate each item
        const payload = JSON.parse(request.payload.toString('utf-8')) as DataRequest;
        this.validateInput(payload);

        results.push(`Processed: ${payload.query}`);
        console.log(`  Processed item ${requestCount}`);

        await this.delay(30);
      }

      const response = {
        totalProcessed: requestCount,
        results: results.slice(0, 10), // Return first 10 as sample
      };

      return {
        payload: Buffer.from(JSON.stringify(response), 'utf-8'),
        metadata: {
          'x-batch-size': String(requestCount),
        },
      };
    } catch (error) {
      console.error('  ✗ Batch processing error:', error);
      throw this.toGrpcError(error);
    }
  }

  /**
   * Validate authentication metadata
   */
  private validateAuth(request: GrpcRequest): void {
    const authToken = request.metadata['authorization'];

    if (!authToken) {
      throw new Error('UNAUTHENTICATED: Missing authorization token');
    }

    if (!authToken.startsWith('Bearer ')) {
      throw new Error('UNAUTHENTICATED: Invalid authorization format');
    }

    // Simulate token validation
    const token = authToken.substring(7);
    if (token.length < 10) {
      throw new Error('UNAUTHENTICATED: Invalid token');
    }
  }

  /**
   * Validate request input
   */
  private validateInput(payload: DataRequest): void {
    if (!payload.userId || payload.userId.trim() === '') {
      throw new Error('INVALID_ARGUMENT: userId is required');
    }

    if (!payload.query || payload.query.trim() === '') {
      throw new Error('INVALID_ARGUMENT: query is required');
    }

    if (payload.query.length > this.MAX_QUERY_LENGTH) {
      throw new Error(
        `INVALID_ARGUMENT: query exceeds maximum length of ${this.MAX_QUERY_LENGTH}`
      );
    }

    if (payload.limit !== undefined && (payload.limit < 1 || payload.limit > this.MAX_LIMIT)) {
      throw new Error(`INVALID_ARGUMENT: limit must be between 1 and ${this.MAX_LIMIT}`);
    }

    // Validate query format (simple check)
    if (payload.query.includes('<script>')) {
      throw new Error('INVALID_ARGUMENT: query contains invalid characters');
    }
  }

  /**
   * Check rate limit
   */
  private checkRateLimit(userId: string): void {
    const allowed = this.rateLimiter.checkLimit(
      userId,
      this.MAX_REQUESTS_PER_MINUTE,
      this.RATE_LIMIT_WINDOW_MS
    );

    if (!allowed) {
      const remaining = this.rateLimiter.getRemainingRequests(
        userId,
        this.MAX_REQUESTS_PER_MINUTE,
        this.RATE_LIMIT_WINDOW_MS
      );

      throw new Error(
        `RESOURCE_EXHAUSTED: Rate limit exceeded (${this.MAX_REQUESTS_PER_MINUTE} requests per minute). Try again later.`
      );
    }
  }

  /**
   * Check user permissions
   */
  private checkPermissions(userId: string, query: string): void {
    // Simulate permission check
    if (query.toLowerCase().includes('admin') && !userId.startsWith('admin-')) {
      throw new Error('PERMISSION_DENIED: Insufficient permissions for admin queries');
    }

    // Simulate sensitive data access
    if (query.toLowerCase().includes('sensitive') && userId.startsWith('guest-')) {
      throw new Error('PERMISSION_DENIED: Guest users cannot access sensitive data');
    }
  }

  /**
   * Convert application errors to gRPC errors with proper status codes
   */
  private toGrpcError(error: unknown): Error {
    if (error instanceof Error) {
      const message = error.message;

      // Extract status code from message prefix
      const match = message.match(/^([A-Z_]+):\s*(.+)$/);
      if (match) {
        const [, code, details] = match;
        return new Error(`gRPC ${code}: ${details}`);
      }

      // Default to INTERNAL for unstructured errors
      return new Error(`gRPC INTERNAL: ${message}`);
    }

    return new Error('gRPC INTERNAL: Unknown error');
  }

  /**
   * Utility: Delay for ms milliseconds
   */
  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

/**
 * Client: Test error handling scenarios
 */
async function runClient() {
  console.log('=== Error Handling Examples ===\n');

  const service = new DataServiceHandler();

  // Test 1: Missing authentication
  console.log('Test 1: Missing Authentication');
  console.log('--------------------------------');
  try {
    await service.handleRequest({
      serviceName: 'data.v1.DataService',
      methodName: 'GetData',
      payload: Buffer.from(JSON.stringify({ userId: 'user1', query: 'test' }), 'utf-8'),
      metadata: {}, // No auth token
    });
    console.log('✗ Should have failed\n');
  } catch (error) {
    console.log(`✓ Expected error: ${error}\n`);
  }

  // Test 2: Invalid input
  console.log('Test 2: Invalid Input');
  console.log('---------------------');
  try {
    await service.handleRequest({
      serviceName: 'data.v1.DataService',
      methodName: 'GetData',
      payload: Buffer.from(JSON.stringify({ userId: '', query: 'test' }), 'utf-8'),
      metadata: { authorization: 'Bearer valid-token-12345' },
    });
    console.log('✗ Should have failed\n');
  } catch (error) {
    console.log(`✓ Expected error: ${error}\n`);
  }

  // Test 3: Permission denied
  console.log('Test 3: Permission Denied');
  console.log('-------------------------');
  try {
    await service.handleRequest({
      serviceName: 'data.v1.DataService',
      methodName: 'GetData',
      payload: Buffer.from(
        JSON.stringify({ userId: 'guest-user1', query: 'sensitive data' }),
        'utf-8'
      ),
      metadata: { authorization: 'Bearer valid-token-12345' },
    });
    console.log('✗ Should have failed\n');
  } catch (error) {
    console.log(`✓ Expected error: ${error}\n`);
  }

  // Test 4: Rate limiting
  console.log('Test 4: Rate Limiting');
  console.log('---------------------');
  try {
    const userId = 'user-rate-test';
    // Make 11 rapid requests (limit is 10 per minute)
    for (let i = 0; i < 11; i++) {
      await service.handleRequest({
        serviceName: 'data.v1.DataService',
        methodName: 'GetData',
        payload: Buffer.from(JSON.stringify({ userId, query: `test ${i}` }), 'utf-8'),
        metadata: { authorization: 'Bearer valid-token-12345' },
      });
      console.log(`  Request ${i + 1} succeeded`);
    }
    console.log('✗ Should have been rate limited\n');
  } catch (error) {
    console.log(`✓ Rate limit enforced: ${error}\n`);
  }

  // Test 5: Mid-stream error
  console.log('Test 5: Mid-stream Error');
  console.log('------------------------');
  try {
    await service.handleServerStream({
      serviceName: 'data.v1.DataService',
      methodName: 'StreamData',
      payload: Buffer.from(
        JSON.stringify({ userId: 'user2', query: 'stream test', limit: 10 }),
        'utf-8'
      ),
      metadata: { authorization: 'Bearer valid-token-12345' },
    });
    console.log('✗ Should have failed mid-stream\n');
  } catch (error) {
    console.log(`✓ Mid-stream error handled: ${error}\n`);
  }

  // Test 6: Successful request
  console.log('Test 6: Successful Request');
  console.log('--------------------------');
  try {
    const response = await service.handleRequest({
      serviceName: 'data.v1.DataService',
      methodName: 'GetData',
      payload: Buffer.from(
        JSON.stringify({ userId: 'user3', query: 'valid query' }),
        'utf-8'
      ),
      metadata: { authorization: 'Bearer valid-token-12345' },
    });

    const data = JSON.parse(response.payload.toString('utf-8')) as DataResponse;
    console.log(`✓ Success: ${data.result}`);
    console.log(`  Remaining requests: ${response.metadata?.['x-rate-limit-remaining']}\n`);
  } catch (error) {
    console.log(`✗ Unexpected error: ${error}\n`);
  }

  console.log('=== Demo Complete ===');
}

/**
 * Main entry point
 */
async function main() {
  await runClient();
}

// Run if executed directly
if (require.main === module) {
  main().catch((error) => {
    console.error('Error:', error);
    process.exit(1);
  });
}

export { DataServiceHandler, RateLimiter, GrpcStatus };

File: examples/python/grpc/error_handling.py
"""Error Handling gRPC Example - Robust Streaming Services.

This example demonstrates proper error handling, timeouts, and rate limiting
in gRPC streaming services.

Use case: Resilient services, resource management, graceful degradation

Run:
    python examples/python/grpc/error_handling.py
"""

from __future__ import annotations

import asyncio
import json
import time
from typing import TYPE_CHECKING

from spikard.grpc import GrpcHandler, GrpcRequest, GrpcResponse

if TYPE_CHECKING:
    from collections.abc import AsyncGenerator, AsyncIterator


class RateLimitedHandler(GrpcHandler):
    """Handler with rate limiting to prevent resource exhaustion."""

    def __init__(self, max_requests_per_second: int = 10) -> None:
        """Initialize rate limiter.

        Args:
            max_requests_per_second: Maximum requests allowed per second
        """
        self.max_requests = max_requests_per_second
        self.request_timestamps: list[float] = []

    async def handle_server_stream(self, request: GrpcRequest) -> AsyncGenerator[GrpcResponse]:
        """Server streaming with rate limiting.

        Raises:
            Exception: RESOURCE_EXHAUSTED if rate limit exceeded
        """
        current_time = time.time()

        # Clean old timestamps (older than 1 second)
        self.request_timestamps = [ts for ts in self.request_timestamps if current_time - ts < 1.0]

        # Check rate limit
        if len(self.request_timestamps) >= self.max_requests:
            error_data = {
                "error": "RESOURCE_EXHAUSTED",
                "message": f"Rate limit exceeded: {self.max_requests} requests/second",
                "retry_after_seconds": 1,
            }
            raise Exception(json.dumps(error_data))

        # Record this request
        self.request_timestamps.append(current_time)

        # Process normally
        req_data = json.loads(request.payload)
        count = req_data.get("count", 5)

        for i in range(count):
            resp_data = {"index": i, "data": f"message_{i}"}
            yield GrpcResponse(payload=json.dumps(resp_data).encode())
            await asyncio.sleep(0.1)


class TimeoutHandler(GrpcHandler):
    """Handler demonstrating timeout handling in streaming RPCs."""

    async def handle_server_stream(self, request: GrpcRequest) -> AsyncGenerator[GrpcResponse]:
        """Server streaming with timeout enforcement.

        Raises:
            Exception: DEADLINE_EXCEEDED if operation takes too long
        """
        req_data = json.loads(request.payload)
        timeout_seconds = req_data.get("timeout_seconds", 5)
        delay_ms = req_data.get("delay_ms", 100)

        start_time = time.time()

        for i in range(100):  # Could stream many items
            # Check timeout before each operation
            elapsed = time.time() - start_time
            if elapsed > timeout_seconds:
                error_data = {
                    "error": "DEADLINE_EXCEEDED",
                    "message": f"Operation exceeded timeout of {timeout_seconds}s",
                    "items_processed": i,
                }
                raise Exception(json.dumps(error_data))

            resp_data = {"index": i, "timestamp": time.time()}
            yield GrpcResponse(payload=json.dumps(resp_data).encode())

            await asyncio.sleep(delay_ms / 1000)


class ValidationHandler(GrpcHandler):
    """Handler demonstrating input validation and error responses."""

    async def handle_client_stream(self, request_stream: AsyncIterator[GrpcRequest]) -> GrpcResponse:
        """Client streaming with input validation.

        Raises:
            Exception: INVALID_ARGUMENT for validation failures
        """
        valid_items = []
        item_count = 0

        async for request in request_stream:
            item_data = json.loads(request.payload)
            item_count += 1

            # Validate required fields
            if "id" not in item_data:
                error_data = {
                    "error": "INVALID_ARGUMENT",
                    "message": f"Missing required field 'id' in item #{item_count}",
                    "field": "id",
                }
                raise Exception(json.dumps(error_data))

            # Validate data types
            if not isinstance(item_data.get("value"), (int, float)):
                error_data = {
                    "error": "INVALID_ARGUMENT",
                    "message": f"Field 'value' must be numeric in item #{item_count}",
                    "field": "value",
                    "received_type": type(item_data.get("value")).__name__,
                }
                raise Exception(json.dumps(error_data))

            # Validate ranges
            if item_data.get("value", 0) < 0:
                error_data = {
                    "error": "INVALID_ARGUMENT",
                    "message": f"Field 'value' must be non-negative in item #{item_count}",
                    "field": "value",
                    "received_value": item_data.get("value"),
                }
                raise Exception(json.dumps(error_data))

            valid_items.append(item_data)

        result = {
            "status": "success",
            "valid_items": len(valid_items),
            "total_value": sum(item.get("value", 0) for item in valid_items),
        }

        return GrpcResponse(payload=json.dumps(result).encode())


class PermissionHandler(GrpcHandler):
    """Handler demonstrating authentication and authorization checks."""

    def __init__(self) -> None:
        """Initialize with user permissions."""
        self.user_permissions = {
            "admin": ["read", "write", "delete"],
            "user": ["read", "write"],
            "guest": ["read"],
        }

    async def handle_server_stream(self, request: GrpcRequest) -> AsyncGenerator[GrpcResponse]:
        """Server streaming with permission checks.

        Raises:
            Exception: UNAUTHENTICATED if no auth token
            Exception: PERMISSION_DENIED if insufficient permissions
        """
        # Check authentication
        auth_token = request.metadata.get("authorization")
        if not auth_token:
            error_data = {
                "error": "UNAUTHENTICATED",
                "message": "Missing authorization token",
                "required_header": "authorization",
            }
            raise Exception(json.dumps(error_data))

        # Extract user role from token (simplified)
        user_role = auth_token.replace("Bearer ", "").strip()

        # Check authorization
        req_data = json.loads(request.payload)
        required_permission = req_data.get("required_permission", "read")

        user_perms = self.user_permissions.get(user_role, [])
        if required_permission not in user_perms:
            error_data = {
                "error": "PERMISSION_DENIED",
                "message": f"User role '{user_role}' lacks permission '{required_permission}'",
                "user_role": user_role,
                "required_permission": required_permission,
                "available_permissions": user_perms,
            }
            raise Exception(json.dumps(error_data))

        # Stream results
        for i in range(5):
            resp_data = {"index": i, "data": f"secure_data_{i}"}
            yield GrpcResponse(payload=json.dumps(resp_data).encode())
            await asyncio.sleep(0.1)


class MidStreamErrorHandler(GrpcHandler):
    """Handler demonstrating mid-stream error handling."""

    async def handle_server_stream(self, request: GrpcRequest) -> AsyncGenerator[GrpcResponse]:
        """Server streaming that may encounter errors mid-stream.

        Yields partial results before raising error.
        """
        req_data = json.loads(request.payload)
        total_items = req_data.get("count", 10)
        error_at = req_data.get("error_at_index", -1)

        for i in range(total_items):
            # Simulate error condition
            if i == error_at:
                error_data = {
                    "error": "INTERNAL",
                    "message": f"Processing failed at item {i}",
                    "items_successfully_processed": i,
                }
                raise Exception(json.dumps(error_data))

            resp_data = {"index": i, "data": f"item_{i}"}
            yield GrpcResponse(payload=json.dumps(resp_data).encode())

            await asyncio.sleep(0.05)


async def simulate_valid_items() -> AsyncIterator[GrpcRequest]:
    """Simulate valid items for validation test."""
    items = [
        {"id": "item1", "value": 100},
        {"id": "item2", "value": 200},
        {"id": "item3", "value": 150},
    ]

    for item in items:
        yield GrpcRequest(
            service_name="validation.v1.ValidationService",
            method_name="ValidateItems",
            payload=json.dumps(item).encode(),
            metadata={},
        )


async def simulate_invalid_items() -> AsyncIterator[GrpcRequest]:
    """Simulate invalid items to trigger validation error."""
    items = [
        {"id": "item1", "value": 100},
        {"value": 200},  # Missing 'id' - should fail
    ]

    for item in items:
        yield GrpcRequest(
            service_name="validation.v1.ValidationService",
            method_name="ValidateItems",
            payload=json.dumps(item).encode(),
            metadata={},
        )


async def example_error_handling() -> None:
    """Demonstrate error handling patterns."""
    # Example 1: Rate limiting
    rate_limiter = RateLimitedHandler(max_requests_per_second=3)

    request = GrpcRequest(
        service_name="api.v1.ApiService",
        method_name="StreamData",
        payload=json.dumps({"count": 5}).encode(),
        metadata={},
    )

    try:
        async for response in rate_limiter.handle_server_stream(request):
            json.loads(response.payload)
    except Exception:
        pass

    # Example 2: Validation
    validator = ValidationHandler()

    try:
        valid_stream = simulate_valid_items()
        await validator.handle_client_stream(valid_stream)
    except Exception:
        pass

    try:
        invalid_stream = simulate_invalid_items()
        await validator.handle_client_stream(invalid_stream)
    except Exception as e:
        json.loads(str(e))

    # Example 3: Permissions
    perm_handler = PermissionHandler()

    # Authorized request
    auth_request = GrpcRequest(
        service_name="secure.v1.SecureService",
        method_name="GetSecureData",
        payload=json.dumps({"required_permission": "read"}).encode(),
        metadata={"authorization": "Bearer user"},
    )

    try:
        async for response in perm_handler.handle_server_stream(auth_request):
            json.loads(response.payload)
    except Exception as e:
        json.loads(str(e))

    # Example 4: Mid-stream error
    error_handler = MidStreamErrorHandler()

    error_request = GrpcRequest(
        service_name="data.v1.DataService",
        method_name="StreamData",
        payload=json.dumps({"count": 10, "error_at_index": 5}).encode(),
        metadata={},
    )

    try:
        async for response in error_handler.handle_server_stream(error_request):
            json.loads(response.payload)
            # Partial results received before error
    except Exception as e:
        json.loads(str(e))


if __name__ == "__main__":
    # Run examples
    asyncio.run(example_error_handling())
